File: Top Interview 150/Hard/MaxPointsOnALine.py
from collections import defaultdict

class Solution(object):
    def maxPoints(self, points):
        """
        :type points: List[List[int]]
        :rtype: int
        """
        if len(points) == 1:
            return 1
        
        n = len(points)
        longest_line = 0

        for i in range(n):
            slope_map = {}
            for j in range(i + 1, n):
                (a, b), (c, d) = points[i], points[j]
                if a == c:
                    slope_map[float('inf')] = slope_map.get(float('inf'), 1) + 1
                elif b == d:
                    slope_map[0] = slope_map.get(0, 1) + 1
                else:
                    slope = (a - c) / (b - d)
                    slope_map[slope] = slope_map.get(slope, 1) + 1
            if slope_map:
                longest_line = max(max(slope_map.values()), longest_line)

        return longest_line


if __name__ == "__main__":
    sol = Solution()
    # points = [[1,1],[3,2],[5,3],[4,1],[2,3],[1,4]]
    points = [[9,-25],[-4,1],[-1,5],[-7,7]]
    print(sol.maxPoints(points))

File: Top Interview 150/Hard/MaxPointsOnALine.java
import java.util.HashMap;

public class MaxPointsOnALine {
    public int maxPoints(int[][] points) {
        int n = points.length;
        if (n < 3) return n;

        
        HashMap<Double, Integer> slopeMap = new HashMap<>(n, 0.75f);
        int longestLine = 0;
        int maxLength = n;
        double slope;
        
        for (int i = 0; i < n; ++i) {
            int x1 = points[i][0];
            int y1 = points[i][1];
            for (int j = i + 1; j < n; ++j) {
                int x2 = points[j][0];
                int y2 = points[j][1];
                if (x1 == x2)
                    slope = Float.POSITIVE_INFINITY;
                else if (y1 == y2)
                    slope = 0f;
                else
                    slope = (float) (y1 - y2) / (x1 - x2);
                
                int count = slopeMap.getOrDefault(slope, 1) + 1;
                slopeMap.put(slope, count);
                longestLine = Math.max(longestLine, count);
            }
            if (longestLine == maxLength--)
                return longestLine;
            slopeMap.clear();
        }
        return longestLine;
    }
    
    public static void main (String[] args) {
        MaxPointsOnALine sol = new MaxPointsOnALine();
        // int[][] points = {{1,1},{3,2},{5,3},{4,1},{2,3},{1,4}};
        // int[][] points = {{9,-25},{-4,1},{-1,5},{-7,7}};
        // int[][] points = {{2,3},{3,3},{-5,3}};
        int[][] points = {{1,1},{3,2},{5,3},{4,1},{2,3},{1,4}};
        System.out.println(sol.maxPoints(points));
    }
}